src/test/scala/io/qbeast/spark/utils/QueryExecutorTest.scala

```diff
@@ -0,0 +1,140 @@
+
+  behavior of "QueryExecutor"
+
+  it should "find all sample files" in withSparkAndTmpDir((spark, tmpdir) => {
+    val source = createDF(1000, spark)
+
+    writeTestData(source.toDF(), Seq("a", "c"), 10, tmpdir)
+
+    val deltaLog = DeltaLog.forTable(spark, tmpdir)
+    val qbeastSnapshot = DeltaQbeastSnapshot(deltaLog.snapshot)
+    val filters = Seq.empty
+
+    val querySpec = new QuerySpecBuilder(filters)
+    val queryExecutor = new QueryExecutor(querySpec, qbeastSnapshot)
+
+    val allDeltaFiles = deltaLog.snapshot.allFiles.collect()
+    val allFiles = allDeltaFiles.map(a => QbeastFile(a.path, a.tags))
+
+    val matchFiles = queryExecutor.execute(allFiles)
+
+    val diff = (allFiles.toSet -- matchFiles.toSet)
+
+    diff.size shouldBe 0
+    matchFiles.size shouldBe allFiles.length
+    matchFiles.toSet shouldBe allFiles.toSet
+
+  })
+
+  it should "filter the files according to the weight" in withSparkAndTmpDir((spark, tmpdir) => {
+    val source = createDF(1000, spark)
+
+    writeTestData(source.toDF(), Seq("a", "c"), 10, tmpdir)
+
+    val deltaLog = DeltaLog.forTable(spark, tmpdir)
+    val qbeastSnapshot = DeltaQbeastSnapshot(deltaLog.snapshot)
+
+    def weightFilters(weightRange: WeightRange): Expression = {
+      val qbeast_hash = new QbeastMurmur3Hash(Seq(new Column("a").expr, new Column("c").expr))
+      val lessThan = LessThan(qbeast_hash, Literal(weightRange.to.value))
+      val greaterThanOrEqual = GreaterThanOrEqual(qbeast_hash, Literal(weightRange.from.value))
+      And(lessThan, greaterThanOrEqual)
+    }
+    val filters = Seq(weightFilters(WeightRange(Weight.MinValue, Weight(0.001))))
+    val querySpec = new QuerySpecBuilder(filters)
+    val queryExecutor = new QueryExecutor(querySpec, qbeastSnapshot)
+
+    val allDeltaFiles = deltaLog.snapshot.allFiles.collect()
+    val allFiles = allDeltaFiles.map(a => QbeastFile(a.path, a.tags))
+
+    val matchFiles = queryExecutor.execute(allFiles)
+
+    matchFiles.size shouldBe <(allFiles.length)
+    matchFiles.foreach(file => allFiles should contain(file))
+
+  })
+
+  it should "filter the files according to the space" in withSparkAndTmpDir((spark, tmpdir) => {
+    val source = createDF(1000, spark)
+
+    writeTestData(source.toDF(), Seq("a", "c"), 10, tmpdir)
+
+    val deltaLog = DeltaLog.forTable(spark, tmpdir)
+    val qbeastSnapshot = DeltaQbeastSnapshot(deltaLog.snapshot)
+
+    val filters = Seq(expr("a >= 2 and a < 10").expr)
+    val querySpec = new QuerySpecBuilder(filters)
+    val queryExecutor = new QueryExecutor(querySpec, qbeastSnapshot)
+
+    val allDeltaFiles = deltaLog.snapshot.allFiles.collect()
+    val allFiles = allDeltaFiles.map(a => QbeastFile(a.path, a.tags))
+
+    val matchFiles = queryExecutor.execute(allFiles)
+
+    matchFiles.size shouldBe <(allFiles.length)
+    matchFiles.foreach(file => allFiles should contain(file))
+
+  })
+
+  it should "find all files in different revisions" in withSparkAndTmpDir((spark, tmpdir) => {
```

Comment:
I don't understand how to this works. The function writeTestData is set to overwrite the dataset, so it should delete the old one, therefore we should have only 1 revision and only the last files. How is it possible this test was even working?

Reply:
Thanks for noticing. I think it's a problem with the Metadata update, I will check it

Reply:
Ok, have seeing the changes now. I will take a look to the lifecycle of revision.

I agree on the revision ID, and of course, overwrite should drop all previous configuration. Delta maintains the schema and everything, but better to drop Qbeast keys before updating.  
In any case, RevisionChanges should notice the overwrite mode as well and set the ID to 1.